# Question 8: Does weather have any effect on headway or on-time performance? To help answer this question, the file bna_2022.csv contains historical weather data recorded at Nashville International Airport.

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

In [2]:
headway = pd.read_csv('../Data/Headway Data.csv')

In [3]:
weather = pd.read_csv('../Data/bna_weather.csv')

stripped hour from data column to merge the two datasets

In [4]:
headway['NEW_DATE']=headway.DATE
headway['NEW_DATE']= pd.to_datetime(headway.NEW_DATE)
headway['HOUR']=headway.SCHEDULED_TIME.str.slice(0,2).astype(str)

In [5]:
weather['HOUR']=weather.Date.str.slice(11,13).astype(str)
weather['NEW_DATE']=weather.Date.str.slice(0,10).astype(str)
weather['NEW_DATE']= pd.to_datetime(weather.NEW_DATE)

In [6]:
merge_df = headway.merge(weather, on=['NEW_DATE','HOUR'], how= "left")
merge_df

CALENDAR_ID  SERVICE_ABBR  ADHERENCE_ID        DATE  ROUTE_ABBR  \
0          120211101             1      76447164  2021-11-01           7   
1          120211101             1      76447165  2021-11-01           7   
2          120211101             1      76447166  2021-11-01           7   
3          120211101             1      76447167  2021-11-01          50   
4          120211101             1      76447168  2021-11-01          50   
...              ...           ...           ...         ...         ...   
2210132    120221101             1      91057724  2022-11-01          50   
2210133    120221101             1      91057851  2022-11-01          50   
2210134    120221101             1      91057852  2022-11-01          50   
2210135    120221101             1      91057881  2022-11-01          56   
2210136    120221101             1      91057882  2022-11-01          56   

         BLOCK_ABBR  OPERATOR  TRIP_ID  OVERLOAD_ID ROUTE_DIRECTION_NAME  ...  \
0              1704      2088   297750            0          TO DOWNTOWN  ...   
1              1704      2088   297750            0          TO DOWNTOWN  ...   
2              1704      2088   297750            0          TO DOWNTOWN  ...   
3              1704      2088   297749            0          TO DOWNTOWN  ...   
4              1704      2088   297749            0          TO DOWNTOWN  ...   
...             ...       ...      ...          ...                  ...  ...   
2210132        9302      2355   329980            0          TO DOWNTOWN  ...   
2210133        9950      1880   330003            0          TO DOWNTOWN  ...   
2210134        9950      1880   330003            0          TO DOWNTOWN  ...   
2210135        9975      1922   330011            0          TO DOWNTOWN  ...   
2210136        9975      1922   330011            0          TO DOWNTOWN  ...   

        clds  water_temp  primary_wave_period  primary_wave_height  \
0        BKN         NaN                  NaN                  NaN   
1        BKN         NaN                  NaN                  NaN   
2        BKN         NaN                  NaN                  NaN   
3        SCT         NaN                  NaN                  NaN   
4        SCT         NaN                  NaN                  NaN   
...      ...         ...                  ...                  ...   
2210132  NaN         NaN                  NaN                  NaN   
2210133  NaN         NaN                  NaN                  NaN   
2210134  NaN         NaN                  NaN                  NaN   
2210135  NaN         NaN                  NaN                  NaN   
2210136  NaN         NaN                  NaN                  NaN   

         primary_swell_period primary_swell_height primary_swell_direction  \
0                         NaN                  NaN                     NaN   
1                         NaN                  NaN                     NaN   
2                         NaN                  NaN                     NaN   
3                         NaN                  NaN                     NaN   
4                         NaN                  NaN                     NaN   
...                       ...                  ...                     ...   
2210132                   NaN                  NaN                     NaN   
2210133                   NaN                  NaN                     NaN   
2210134                   NaN                  NaN                     NaN   
2210135                   NaN                  NaN                     NaN   
2210136                   NaN                  NaN                     NaN   

        secondary_swell_period  secondary_swell_height  \
0                          NaN                     NaN   
1                          NaN                     NaN   
2                          NaN                     NaN   
3                          NaN                     NaN   
4                          NaN                     NaN 

In [8]:
merge_df = merge_df.drop_duplicates(['ADHERENCE_ID'])

In [9]:
merge_df['DEV_PERCENT'] = merge_df.HDWY_DEV/merge_df.SCHEDULED_HDWY * 100

In [10]:
def BUNCH_GAP(value):
    if value >= 50:
        return "Gapped"
    elif value <= -50:
        return "Bunched" 
    else:
        return "On Time"

merge_df['BUNCH_GAP'] = merge_df['DEV_PERCENT'].map(BUNCH_GAP)
merge_df


CALENDAR_ID  SERVICE_ABBR  ADHERENCE_ID        DATE  ROUTE_ABBR  \
0          120211101             1      76447164  2021-11-01           7   
1          120211101             1      76447165  2021-11-01           7   
12         120211101             1      76447976  2021-11-01          22   
13         120211101             1      76447977  2021-11-01          22   
14         120211101             1      76447978  2021-11-01          22   
...              ...           ...           ...         ...         ...   
2210120    120221101             1      91056034  2022-11-01           7   
2210121    120221101             1      91056035  2022-11-01           7   
2210129    120221101             1      91057381  2022-11-01          55   
2210133    120221101             1      91057851  2022-11-01          50   
2210135    120221101             1      91057881  2022-11-01          56   

         BLOCK_ABBR  OPERATOR  TRIP_ID  OVERLOAD_ID ROUTE_DIRECTION_NAME  ...  \
0              1704      2088   297750            0          TO DOWNTOWN  ...   
1              1704      2088   297750            0          TO DOWNTOWN  ...   
12             2200      1352   298128            0          TO DOWNTOWN  ...   
13             2200      1352   298128            0          TO DOWNTOWN  ...   
14             2200      1352   298128            0          TO DOWNTOWN  ...   
...             ...       ...      ...          ...                  ...  ...   
2210120         706       375   329459            0        FROM DOWNTOWN  ...   
2210121         706       375   329459            0        FROM DOWNTOWN  ...   
2210129        8604      2374   329899            0          TO DOWNTOWN  ...   
2210133        9950      1880   330003            0          TO DOWNTOWN  ...   
2210135        9975      1922   330011            0          TO DOWNTOWN  ...   

        primary_wave_period  primary_wave_height  primary_swell_period  \
0                       NaN                  NaN                   NaN   
1                       NaN                  NaN                   NaN   
12                      NaN                  NaN                   NaN   
13                      NaN                  NaN                   NaN   
14                      NaN                  NaN                   NaN   
...                     ...                  ...                   ...   
2210120                 NaN                  NaN                   NaN   
2210121                 NaN                  NaN                   NaN   
2210129                 NaN                  NaN                   NaN   
2210133                 NaN                  NaN                   NaN   
2210135                 NaN                  NaN                   NaN   

         primary_swell_height  primary_swell_direction secondary_swell_period  \
0                         NaN                      NaN                    NaN   
1                         NaN                      NaN                    NaN   
12                        NaN                      NaN                    NaN   
13                        NaN                      NaN                    NaN   
14                        NaN                      NaN                    NaN   
...                       ...                      ...                    ...   
2210120                   NaN                      NaN                    NaN   
2210121                   NaN                      NaN                    NaN   
2210129                   NaN                      NaN                    NaN   
2210133                   NaN                      NaN                    NaN   
2210135                   NaN                      NaN                    NaN   

        secondary_swell_height secondary_swell_direction  DEV_PERCENT  \
0                          NaN                       NaN    14.166664   
1                          NaN                       NaN    23.809521   
12                         NaN            

In [11]:
merge_df['BUNCH_GAP'].value_counts(dropna=False)



On Time    1392843
Gapped       86564
Bunched      70014
Name: BUNCH_GAP, dtype: int64

Dropping nulls in dev_percent column

On Time    1350769
Gapped       86564
Bunched      70014

Dropping nulls in scheduled headway column:

On Time    1392843
Gapped       86564
Bunched      70014

On Time    1350769
Gapped       86564
Bunched      70014

Without dropping nulls:

On Time    1825137
Gapped       86564
Bunched      70014


In [ ]:
merge_df['BUNCH_GAP'].value_counts(normalize=True)

In [ ]:
below_freezing=merge_df.loc[merge_df['temp']<32]
below_freezing['BUNCH_GAP'].value_counts()

In [ ]:
below_freezing['BUNCH_GAP'].value_counts(normalize=True)

In [ ]:
above_freezing['BUNCH_GAP'].value_counts()
above_freezing['BUNCH_GAP'].value_counts()

In [ ]:
above_freezing['BUNCH_GAP'].value_counts(normalize = True)

In [ ]:
z

In [ ]:
above_freezing['BUNCH/GAP'].value_counts().plot(kind='bar');

In [ ]:
below_freezing['BUNCH/GAP'].value_counts().plot(kind='bar');

In [ ]:
# exit_df['Exit Reason'].value_counts(normalize = True)
above_freezing['BUNCH/GAP'].value_counts(normalize=True)

In [ ]:
below_freezing['BUNCH/GAP'].value_counts(normalize=True)

In [ ]:
percentiles = [.1, .2, .3, .4, .5, .6, .7, .8, .9]
include = 'all'
above_freezing.mean()

In [ ]:
above_freezing['temp'].value_counts().plot(kind='hist');

In [ ]:
merge_df_b= merge_df_a[['NEW_DATE', 'HOUR','DATE','Date', 'ROUTE_ABBR','TRIP_ID',
       'ROUTE_DIRECTION_NAME', 'SCHEDULED_TIME',
       'ACTUAL_ARRIVAL_TIME', 'ACTUAL_DEPARTURE_TIME', 'ADHERENCE',
       'SCHEDULED_HDWY', 'ACTUAL_HDWY', 'HDWY_DEV',
       'temp', 'heat_index', 'precip_hrly']]

In [ ]:
freezing_rain = below_freezing[['temp','precip_hrly', 'BUNCH/GAP']]
freezing_rain

In [ ]:
#above_freezing=merge_df_b.loc[merge_df_b['temp']>32]

freezing_rain = below_freezing.loc[below_freezing['precip_hrly']>0]
freezing_rain.head()

In [ ]:
below_freezing['precip_hrly'].value_counts()

In [ ]:
freezing_rain['BUNCH/GAP'].value_counts()

In [ ]:
freezing_rain['BUNCH/GAP'].value_counts(normalize=True)

In [ ]:
freezing_rain['BUNCH/GAP'].value_counts().plot(kind='bar')

In [ ]:
merge_df['temp'].plot(kind='hist')
plt.style.use('seaborn')




In [ ]:
# plt.hist(df_2014['Internet_Users_Pct'], bins=8)


plt.hist(merge_df['temp'], bins=6)
plt.show()

In [ ]:

plt.plot(merge_df['temp'])
plt.show()

In [ ]:
merge_df['temp'].value_counts().plot(kind='barh')

In [ ]:
merge_df_b.head()

In [ ]:
merge_df_b.plot.scatter('temp', 'DEV_PERCENT', s=None, c='green')



In [ ]:
below_freezing.corr()
